osu.Game.Tests/Visual/Gameplay/TestSceneBeatmapSkinFallbacks.cs

```diff
@@ -0,0 +1,135 @@
+                Size = actualComponentsContainer.DrawSize,
+                Child = expectedComponentsContainer,
+            };
+
+            Add(expectedComponentsAdjustmentContainer);
+            expectedComponentsAdjustmentContainer.UpdateSubTree();
+            var expectedInfo = expectedComponentsContainer.CreateSkinnableInfo();
+            Remove(expectedComponentsAdjustmentContainer);
+
+            return almostEqual(actualInfo, expectedInfo);
+
+            static bool almostEqual(SkinnableInfo info, SkinnableInfo other) =>
+                other != null
+                && info.Type == other.Type
+                && info.Anchor == other.Anchor
+                && info.Origin == other.Origin
+                && Precision.AlmostEquals(info.Position, other.Position)
+                && Precision.AlmostEquals(info.Scale, other.Scale)
+                && Precision.AlmostEquals(info.Rotation, other.Rotation)
+                && info.Children.SequenceEqual(other.Children, new FuncEqualityComparer<SkinnableInfo>(almostEqual));
+        }
+
+        protected override WorkingBeatmap CreateWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null)
+            => new CustomSkinWorkingBeatmap(beatmap, storyboard, Clock, Audio, currentBeatmapSkin);
+
+        protected override Ruleset CreatePlayerRuleset() => new TestOsuRuleset();
+
+        private class CustomSkinWorkingBeatmap : ClockBackedTestWorkingBeatmap
+        {
+            private readonly ISkin beatmapSkin;
+
+            public CustomSkinWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard, IFrameBasedClock referenceClock, AudioManager audio, ISkin beatmapSkin)
+                : base(beatmap, storyboard, referenceClock, audio)
+            {
+                this.beatmapSkin = beatmapSkin;
+            }
+
+            protected override ISkin GetSkin() => beatmapSkin;
+        }
+
+        private class TestOsuRuleset : OsuRuleset
+        {
+            public override ISkin CreateLegacySkinProvider(ISkinSource source, IBeatmap beatmap) => new TestOsuLegacySkinTransformer(source);
+
+            private class TestOsuLegacySkinTransformer : OsuLegacySkinTransformer
+            {
+                public TestOsuLegacySkinTransformer(ISkinSource source)
+                    : base(source)
+                {
+                }
+
+                public override Drawable GetDrawableComponent(ISkinComponent component)
+                {
+                    var drawable = base.GetDrawableComponent(component);
+                    if (drawable != null)
+                        return drawable;
+
+                    // this isn't really supposed to make a difference from returning null,
+                    // but it appears it does, returning null will skip over falling back to beatmap skin,
+                    // while calling Source.GetDrawableComponent() doesn't.
```

Comment:
I don't get this. Are you saying this is a bug? `OsuRulesetTransformer` does return `null` - are you suggesting that it is never going to use the beatmap skin as a result? Because that's how this reads.

https://github.com/ppy/osu/blob/004798d61d44d88114acfd7c2861378261eab107/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs#L120

Reply:
That's correct, if `LegacyBeatmapSkin.GetDrawableComponent()` implements something for a certain `component` and the transformer returns `null` on that `component`, then the implementation in the beatmap skin will be completely ignored.

And this is currently a hidden bug, until beatmap skins can have skinned HUD components, in which case they'll be silently ignored and the bug would become visible.

And now that I think about it, it sounds like the transformers should just return `Source.GetDrawableComponent` if it's not handling it, while allowing to return `null` if there are no skinned textures for the component (in which case it doesn't make sense to call the source).

Reply:
The transformer's responsibility is to decide what should or should not be seen in the usage scenario. If the correct behaviour is that "unknowns should be provided to the consumer" then yes, the default `null` should be replaced with a fallback call to the transformation source.

Reply:
Which is to say, your inline comment reads very badly. Try rewording it.

And maybe it should explain why we're doing that here rather than just accepting what `base.GetDrawableComponent` is giving us?